pkg/controller/async_backup_operations_controller.go

```diff
@@ -0,0 +1,491 @@
+		copy(*out, *in)
+	}
+}
+
+func (o *operationsForBackup) uploadProgress(backupStore persistence.BackupStore, backupName string) error {
+	if len(o.operations) > 0 {
+		var backupItemOperations *bytes.Buffer
+		backupItemOperations, errs := encodeToJSONGzip(o.operations, "backup item operations list")
+		if errs != nil {
+			return errors.Wrap(errs[0], "error encoding item operations json")
+		}
+		err := backupStore.PutBackupItemOperations(backupName, backupItemOperations)
+		if err != nil {
+			return errors.Wrap(err, "error uploading item operations json")
+		}
+	}
+	o.changesSinceUpdate = false
+	o.errsSinceUpdate = nil
+	return nil
+}
+
+type BackupItemOperationsMap struct {
+	operations map[string]*operationsForBackup
+	opsLock    sync.Mutex
+}
+
+// If backup has changes not yet uploaded, upload them now
+func (m *BackupItemOperationsMap) UpdateForBackup(backupStore persistence.BackupStore, backupName string) error {
+	// lock operations map
+	m.opsLock.Lock()
+	defer m.opsLock.Unlock()
+
+	operations, ok := m.operations[backupName]
+	// if operations for this backup aren't found, or if there are no changes
+	// or errors since last update, do nothing
+	if !ok || (!operations.changesSinceUpdate && len(operations.errsSinceUpdate) == 0) {
+		return nil
+	}
+	if err := operations.uploadProgress(backupStore, backupName); err != nil {
+		return err
+	}
+	return nil
+}
+
+type asyncBackupOperationsReconciler struct {
+	client.Client
+	logger            logrus.FieldLogger
+	clock             clock.Clock
+	frequency         time.Duration
+	itemOperationsMap *BackupItemOperationsMap
+	newPluginManager  func(logger logrus.FieldLogger) clientmgmt.Manager
+	backupStoreGetter persistence.ObjectBackupStoreGetter
+	metrics           *metrics.ServerMetrics
+}
+
+func NewAsyncBackupOperationsReconciler(
+	logger logrus.FieldLogger,
+	client client.Client,
+	frequency time.Duration,
+	newPluginManager func(logrus.FieldLogger) clientmgmt.Manager,
+	backupStoreGetter persistence.ObjectBackupStoreGetter,
+	metrics *metrics.ServerMetrics,
+) (*asyncBackupOperationsReconciler, *BackupItemOperationsMap) {
+	abor := &asyncBackupOperationsReconciler{
+		Client:            client,
+		logger:            logger,
+		clock:             clock.RealClock{},
+		frequency:         frequency,
+		itemOperationsMap: &BackupItemOperationsMap{operations: make(map[string]*operationsForBackup)},
+		newPluginManager:  newPluginManager,
+		backupStoreGetter: backupStoreGetter,
+		metrics:           metrics,
+	}
+	if abor.frequency <= 0 {
+		abor.frequency = defaultAsyncBackupOperationsFrequency
+	}
+	return abor, abor.itemOperationsMap
+}
+
+func (c *asyncBackupOperationsReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	s := kube.NewPeriodicalEnqueueSource(c.logger, mgr.GetClient(), &velerov1api.BackupList{}, c.frequency, kube.PeriodicalEnqueueSourceOption{})
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&velerov1api.Backup{}, builder.WithPredicates(predicate.Funcs{
```

Comment:
Does this mean the controller wants to reconcile Backup CRs with New event happens on them? Do we need to reconcile the newly created backups in this controller?

Reply:
I've modeled this after the GC controller. It's a periodical enqueue source, so the reconciles on all backups every 2 minutes. Ahh, but I see your point here. GC controller also acts on created backups so that newly-synced backups will be deleted if they're expired. We don't need to run reconcile on backup creation here, since the only backups we care about are those waiting for async actions. So it looks like I need to add this:
```
			CreateFunc: func(ue event.CreateEvent) bool {
				return false
			},
```

Reply:
Actually, I can just remove this entire custom predicate and use kube.FalsePredicate{} instead.

Reply:
Fixed